README
ident is a simple Ruby library for querying identd servers (RFC 1413).
For usage see the documentation and and the examples.


CHANGELOG
2010/08/11
version 0.0.3
* fixed a bug where additional whitespace in a response would break
  the code

2009/10/14
version 0.0.2
* fixed a bug in parsing the charset
* added tests

version 0.0.1
* initial release


test/setup.rb
$:.unshift(File.absolute_path("#{__FILE__}../../../lib"))
require "ident"
require 'baretest/use/mocha'


test/suite/lib/ident.rb
require 'baretest'

BareTest.suite("ident") do
  suite "::Response" do
    suite ".from" do
      assert "A response should be built from a string" do
        Ident::Response.from(
                             "0, 0:USERID:SYSTEM:YOURID"
                             ).is_a? Ident::Response::BasicResponse
      end

      assert "A string containing an userid should result in a positive response" do
        Ident::Response.from(
                             "0, 0:USERID:SYSTEM:YOURID"
                             ).is_a? Ident::Response::USERID
      end

      assert "A string containing an error should result in a errornous response" do
        Ident::Response.from(
                             "0, 0:ERROR:NO-USER"
                             ).is_a? Ident::Response::ERROR
      end
    end # .from

    suite "An USERID response" do
      setup do
        @response1 = Ident::Response.from("0, 0:USERID:SYSTEM:YOURID")
        @response2 = Ident::Response.from("0, 0:USERID:SYSTEM-UTF-8:YOURID")
        @response3 = Ident::Response.from("0, 0 : USERID : SYSTEM-UTF-8 : YOURID")
      end

      assert "should contain the userid" do
        equal("YOURID", @response1.userid)
      end

      assert "should contain the operating system" do
        equal("SYSTEM", @response1.os)
      end

      assert "should contain the charset on the machine" do
        equal("UTF-8", @response2.charset)
      end

      assert "should have a default charset" do
        equal("US-ASCII", @response1.charset)
      end

      assert "should not be an error" do
        not @response1.error?
      end

      assert "should strip spaces" do
        equal("YOURID", @response3.userid)
        equal("UTF-8",  @response3.charset)
        equal("SYSTEM", @response3.os)
      end
    end # An USERID response

    suite "An ERROR response" do
      setup do
        @response1 = Ident::Response.from("0, 0:ERROR:INVALID-PORT")
        @response2 = Ident::Response.from("0, 0:ERROR:NO-USER")
        @response3 = Ident::Response.from("0, 0:ERROR:HIDDEN-USER")
        @response4 = Ident::Response.from("0, 0:ERROR:UNKNOWN-ERROR")
        @response5 = Ident::Response.from("0, 0 : ERROR : NO-USER")
      end

      assert "is an error" do
        @response1.error?
      end

      assert "has a type" do
        equal("INVALID-PORT", @response1.type)
      end

      suite "knows its type" do
        assert "if it is an invalid port" do
          @response1.invalid_port? &&
            !@response1.no_user? &&
            !@response1.hidden_user? &&
            !@response1.unknown_error?
        end

        assert "if it is an invalid port" do
          !@response2.invalid_port? &&
            @response2.no_user? &&
            !@response2.hidden_user? &&
            !@response2.unknown_error?
        end

        assert "if it is a hidden user" do
          !@response3.invalid_port? &&
            !@response3.no_user? &&
            @response3.hidden_user? &&
            !@response3.unknown_error?
        end

        assert "if it is an unknown error" do
          not @response4.invalid_port?
          not @response4.no_user?
          not@response4.hidden_user?
          @response4.unknown_error?
        end
      end

      assert "should strip spaces" do
        @response5.no_user?
      end
    end # An ERROR response

    suite "A BasicResponse" do
      setup do
        @response = Ident::Response::BasicResponse.new
      end

      assert "is not an error" do
        not @response.error?
      end

      assert "has empty default values" do
        @response.os.nil? and @response.charset.nil? and @response.userid.nil?
      end
    end # A BasicResponse
  end # ::Response

  suite ".request" do
    assert "should connect to a server" do
      TCPSocket.expects(:new).with("127.0.0.1", 113).returns(StringIO.new)

      Ident.request("127.0.0.1", 12345, 12345)
    end

    assert "should return a new response" do
      io = StringIO.new("0,0:ERROR:UNKNOWN-ERROR")
      io.expects(:puts)
      TCPSocket.expects(:new).with("127.0.0.1", 113). returns(io).at_least_once

      Ident.request("127.0.0.1", 12345, 12345).is_a? Ident::Response::BasicResponse
    end

    assert "should return a new response if the server terminates the connection" do
      TCPSocket.expects(:new).with("127.0.0.1", 113).returns(StringIO.new).at_least_once
      Ident.request("127.0.0.1", 12345, 12345).is_a? Ident::Response::ERROR
    end
  end

  suite "as an instance" do
    assert "should act as a proxy to Ident.request" do
      Ident.expects(:request).with("127.0.0.1", 12345, 6789, 20).returns(true).once

      i = Ident.new
      i.ip       = "127.0.0.1"
      i.outbound = 12345
      i.inbound  = 6789
      i.timeout  = 20

      i.request
    end
  end
end


examples/example.rb
require 'ident'

# Note: none of this examples will work when executed directly, they
# are rather meant for giving an overview how you would use the
# library in your own server applications.

# using the class method Ident.request
begin
  response = Ident.request("127.0.0.1", 12345, 12345)
  # if Ident.request returned an error, userid will be nil. If you want
  # to know what the error was, check Response::ERROR#type or use the
  # methods like `no_user?`
  username = response.userid || "fallback username"
rescue Timeout::Error, Errno::ECONNREFUSED
  # we couldn't connect to an identd, use a fallback username
  username = "fallback username"
end

# using an instance of Ident. basically the same as above, but you can
# set the arguments before invoking the request
begin
  i = Ident.new
  i.ip = "127.0.0.1"
  i.outbound = 12345
  i.inbound  = 12345
  i.timeout  = 5

  response = i.request
  username = response.userid || "fallback username"
rescue Timeout::Error, Errno::ECONNREFUSED
  username = "fallback username"
end

# you can ask for the userid, the os and the charset.


lib/ident/response/error.rb
require "ident/response/basic_response"

class Ident
  module Response
    # This class gives access to the error returned by an identd. It
    # defines methods for every possible kind of error:
    # {invalid_port?}, {no_user?}, {hidden_user?}, {unknown_error?}
    #
    # Note: You should not instantiate this class yourself.
    class ERROR < BasicResponse
      KNOWN_ERRORS.each do |e|
        define_method(error_to_method_name(e)) do
          @type == e
        end
      end

      # @return [String]
      attr_reader :type
      def initialize(type)
        @type = type
      end

      # @see BasicResponse#error?
      def error?; true; end

      # @return [Boolean] Did we specify an invalid port?
      def invalid_port?; @type == "INVALID-PORT"; end

      # @return [Boolean]  Is no user known for the given connection specification?
      def no_user?; @type == "NO-USER"; end

      # @return [Boolean] Does the identd hide information from us?
      def hidden_user?; @type == "HIDDEN-USER"; end

      # @return [Boolean] Did an unknown error occur?
      def unknown_error?; @type == "UNKNOWN-ERROR"; end
    end
  end
end


lib/ident/response/user_id.rb
require "ident/response/basic_response"

class Ident
  module Response
    # This class provides access to information returned about the
    # user belonging to a connection.
    #
    # Note: You should not instantiate this class yourself.
    class USERID < BasicResponse
      # The operating system of the user
      #
      # @return [String]
      attr_reader :os

      # The charset on the system, defaults to US-ASCII
      #
      # @return [String]
      attr_reader :charset

      # The userid
      #
      # @return [String]
      attr_reader :userid

      # @param [String] os The os+charset string
      # @param [String] userid
      def initialize(os, userid)
        parts = os.split('-')
        @os = parts[0]
        @charset = parts[1..-1].join('-')
        @charset = 'US-ASCII' if @charset.empty?
        @userid = userid
      end
    end
  end
end


lib/ident/response/basic_response.rb
class Ident
  module Response
    # This is the blankslate response which ensures that both kind of
    # responses respond to all required methods. You should never have
    # to work with this class directly.
    #
    # @api private
    class BasicResponse
      # Is the response the result of an error?
      #
      # @return [Boolean]
      def error?;  false; end

      # @return [nil]
      def os;      nil;   end
      # @return [nil]
      def charset; nil;   end
      # @return [nil]
      def userid;  nil;   end

      class << self
        # @return [String]
        def error_to_method_name(error)
          (error.downcase.tr('-', '_') + "?").to_sym
        end
        private :error_to_method_name
      end

      public
      KNOWN_ERRORS.each do |e|
        define_method(error_to_method_name(e)) do
          false
        end
      end
    end
  end
end


lib/ident.rb
require 'socket'
require 'timeout'

class Ident
  KNOWN_ERRORS = ["INVALID-PORT" "NO-USER" "HIDDEN-USER" "UNKNOWN-ERROR"]
  require "ident/response"

  # @return [String]
  attr_accessor :ip

  # @return [Number]
  attr_accessor :outbound

  # @return [Number]
  attr_accessor :inbound

  # @return [Number]
  attr_accessor :timeout

  # @return [Response::USERID, Response::ERROR]
  attr_reader :response

  # @param (see Ident.request)
  # @see Ident.request
  def initialize(ip = nil, outbound = nil, inbound = nil, timeout_after = 10)
    @ip, @outbound, @inbound, @timeout = ip, outbound, inbound, timeout_after
    @response = nil
  end


  # Calls {Ident.request} using the attributes supplied to the
  # constructor. The result can be accessed using {#response}
  #
  # @return [void]
  # @see Ident.request
  def request
    @response = self.class.request(@ip, @outbound, @inbound, @timeout)
  end

  # This method connects to an identd at `ip` and queries information
  # for the connection `outbound`->`inbound` where `outbound` is the
  # port a connection is coming from (usually anything >1024) and
  # `inbound` is the port connected to, the port your service is
  # listening on (e.g. 6667 in the case of an IRCd)
  #
  #
  # @example user A connects to our server on port 23, coming from 123.123.123.123 on port 6789
  #  Ident.request('123.123.123.123', 6789, 23)
  #
  # @param [String]  ip            The IP of the ident server
  # @param [Integer] outbound      The port from which the connection is coming
  # @param [Integer] inbound       The port to which was connected
  # @param [Integer] timeout_after The timeout, defaults to 10 seconds
  #
  # @raise [Timeout::Error] In case of a timeout
  # @raise [Errno::ECONNREFUSED] If no identd was running at the destination host
  #
  # @return [Response::USERID] if we successfully queried the ident server
  # @return [Response::ERROR]  in case of an error
  def self.request(ip, outbound, inbound, timeout_after = 10)
    r = nil
    timeout(timeout_after) do
      t = TCPSocket.new(ip, 113)
      t.puts [outbound, inbound].join(', ')
      r = t.gets
      t.close
    end

    # in case the identd just kills our connection
    r ||= "#{outbound}, #{inbound}:ERROR:UNKNOWN-ERROR"

    r.chomp!
    Response.from(r)
  end
end



